下载B站视频超时 在我的VPS上经常出现

panic: Get https://interface.bilibili.com/v2/playurl?appkey=84956560bc028eb7&cid=49605674&otype=json&qn=15&quality=15&type=&sign=6f15acbe924791f0eb4b7c1b042ce0cd&utoken=YWlkPTI4NjQ1NTM3JmNpZD00OTYwNTY3NCZmcm9tPXBjJm1pZD0zNTE2NDgyMDQmb3duZXI9MCZzdmlwPTAmdHM9MTUzMzUzNDU1OCZ2aXA9MCZmY3M9MjBjZTdiMDcwYzhiNTk5ZDg3ZDJjYTRmYjhiMWVmMTU=: net/http: TLS handshake timeout

goroutine 1 [running]:
github.com/iawia002/annie/request.Request(0x71e14a, 0x3, 0xc4201d48c0, 0x13e, 0x0, 0x0, 0xc42018f450, 0x600000001)
	/Users/mac/iawia002/go/src/github.com/iawia002/annie/request/request.go:98 +0xaf7
github.com/iawia002/annie/request.Get(0xc4201d48c0, 0x13e, 0x72947b, 0x18, 0xc42018f450, 0x0, 0x0)
	/Users/mac/iawia002/go/src/github.com/iawia002/annie/request/request.go:129 +0xb7
github.com/iawia002/annie/extractors/bilibili.bilibiliDownload(0x7ffe220fe891, 0x29, 0x0, 0x1, 0xc4201de820, 0xd, 0xc4201de648, 0x8, 0xc4201de8b8, 0x8, ...)
	/Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/bilibili/bilibili.go:221 +0x21e
github.com/iawia002/annie/extractors/bilibili.Download(0x7ffe220fe891, 0x29)
	/Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/bilibili/bilibili.go:152 +0x862
main.download(0x7ffe220fe891, 0x29)
	/Users/mac/iawia002/go/src/github.com/iawia002/annie/main.go:64 +0x82c
main.main()
	/Users/mac/iawia002/go/src/github.com/iawia002/annie/main.go:133 +0xe0

--- a/README.md
+++ b/README.md
@@ -32,6 +32,7 @@
   * [Multiple inputs](#multiple-inputs)
   * [Resume a download](#resume-a-download)
   * [Cookies](#cookies)
+  * [Auto retry](#auto-retry)
   * [Proxy](#proxy)
   * [Multi-Thread](#multi-thread)
   * [Short link](#short-link)
@@ -266,6 +267,10 @@ $ annie -F ~/Desktop/u.txt
 
 A temporary `.download` file is kept in the output directory. If `annie` is ran with the same arguments, then the download progress will resume from the last session.
 
+### Auto retry
+
+annie will auto retry when the download failed, you can specify the retry times by `-retry` option (default is 100).
+
 ### Cookies
 
 Cookies can be provided to `annie` with the `-c` option if they are required for accessing the video.
@@ -460,6 +465,8 @@ $ annie -j https://www.bilibili.com/video/av20203945
     	SOCKS5 proxy
   -x string
     	HTTP proxy
+  -retry int
+    	How many times to retry when the download failed (default 100)
 ```
 
 #### Playlist:

--- a/config/config.go
+++ b/config/config.go
@@ -39,6 +39,8 @@ var (
 	Caption bool
 	// Ccode youku ccode
 	Ccode string
+	// RetryTimes how many times to retry when the download failed
+	RetryTimes int
 )
 
 // FakeHeaders fake http headers

--- a/downloader/downloader.go
+++ b/downloader/downloader.go
@@ -87,7 +87,7 @@ func Caption(url, refer, fileName, ext string) {
 
 func writeFile(
 	url string, file *os.File, headers map[string]string, bar *pb.ProgressBar,
-) {
+) (int64, error) {
 	res := request.Request("GET", url, nil, headers)
 	if res.StatusCode >= 400 {
 		red := color.New(color.FgRed)
@@ -98,10 +98,11 @@ func writeFile(
 	writer := io.MultiWriter(file, bar)
 	// Note that io.Copy reads 32kb(maximum) from input and writes them to output, then repeats.
 	// So don't worry about memory.
-	_, copyErr := io.Copy(writer, res.Body)
+	written, copyErr := io.Copy(writer, res.Body)
 	if copyErr != nil {
-		log.Fatal(fmt.Sprintf("file copy error: %s", copyErr))
+		return written, fmt.Errorf("file copy error: %s", copyErr)
 	}
+	return written, nil
 }
 
 // Save save url file
@@ -165,11 +166,33 @@ func Save(
 		for ; i <= chunk; i++ {
 			end = start + chunkSize - 1
 			headers["Range"] = fmt.Sprintf("bytes=%d-%d", start, end)
-			writeFile(urlData.URL, file, headers, bar)
+			temp := start
+			for i := 0; ; i++ {
+				written, err := writeFile(urlData.URL, file, headers, bar)
+				if err == nil {
+					break
+				} else if i+1 >= config.RetryTimes {
+					log.Fatal(err)
+				}
+				temp += written
+				headers["Range"] = fmt.Sprintf("bytes=%d-%d", temp, end)
+				time.Sleep(1 * time.Second)
+			}
 			start = end + 1
 		}
 	} else {
-		writeFile(urlData.URL, file, headers, bar)
+		temp := tempFileSize
+		for i := 0; ; i++ {
+			written, err := writeFile(urlData.URL, file, headers, bar)
+			if err == nil {
+				break
+			} else if i+1 >= config.RetryTimes {
+				log.Fatal(err)
+			}
+			temp += written
+			headers["Range"] = fmt.Sprintf("bytes=%d-", temp)
+			time.Sleep(1 * time.Second)
+		}
 	}
 	// close and rename temp file at the end of this function
 	defer func() {
@@ -309,7 +332,6 @@ func (v VideoData) Download(refer string) {
 			defer wgp.Done()
 			Save(url, refer, fileName, bar)
 		}(url, refer, partFileName, bar)
-
 	}
 	wgp.Wait()
 	bar.Finish()

--- a/main.go
+++ b/main.go
@@ -38,6 +38,9 @@ func init() {
 	)
 	flag.BoolVar(&config.Caption, "C", false, "Download captions")
 	flag.StringVar(&config.Ccode, "ccode", "0808", "Youku ccode")
+	flag.IntVar(
+		&config.RetryTimes, "retry", 100, "How many times to retry when the download failed",
+	)
 }
 
 func download(videoURL string) {

--- a/request/request.go
+++ b/request/request.go
@@ -84,18 +84,18 @@ func Request(
 	if config.Refer != "" {
 		req.Header.Set("Referer", config.Refer)
 	}
-	retryTimes := 3
 	var (
 		res          *http.Response
 		requestError error
 	)
-	for i := 0; i < retryTimes; i++ {
+	for i := 0; ; i++ {
 		res, requestError = client.Do(req)
-		if requestError == nil {
+		if requestError == nil && res.StatusCode < 400 {
 			break
-		}
-		if requestError != nil && i+1 == retryTimes {
-			panic(requestError)
+		} else if i+1 >= config.RetryTimes {
+			panic(fmt.Sprintf(
+				"request error: %s HTTP %d", requestError, res.StatusCode,
+			))
 		}
 		time.Sleep(1 * time.Second)
 	}

